### PR TITLE
Allow international addresses

### PR DESCRIPTION
**Pull request by thusoy**
[International email addresses](https://en.wikipedia.org/wiki/International_email) are becoming more and more popular, validators shouldn't prevent that. In addition to the unit tests I've also verified this works with all the example emails from the wikipedia article.

### CODE REVIEW

**Comment by scottgonzalez** [comment]
In the real world, internationalization doesn’t work so well, and therefore does not meet the criteria of being sane. The Wikipedia article you linked to even mentions that the UTF-8 encoded addresses have support issues.

There first email validation I wrote fully supported internationalization, along with many other spec-valid rules that just don’t make sense in the real world using real (old) infrastructure to support the Internet. That’s why I created this module to reflect reality.

**Comment by thusoy** [comment]
I agree there's a lot of other rules in the spec that it doesn't make sense to allow, like quoted parts and comments. Those have never been used much and doesn't make sense to allow. Validators like this one is exactly the kind of support problems international email encounters, and while the other rules are not used, international email use is growing.

Another problem is that a lot of projects used the old validation rule you wrote, and cannot upgrade to this one since it's not compatible with stuff that is already in their databases.

**Comment by scottgonzalez** [comment]
How about we expose a third method `isUnicodeEmail` (or similar name)? Or perhaps even change the existing one to `isAsciiEmail`.

**Comment by thusoy** [comment]
I think that's a fair compromise, I'll amend to accommodate that.

**Comment by thusoy** [comment]
Okay, I added `isAsciiEmail/isNotAsciiEmail` with an extra check to disallow any punycoded labels in the domain. The tests are a bit repetitive as I followed the same pattern that was already used, but I think it's fine. Let me know what you think.

File: index.js [hunk 1]
```diff
@@ -1,8 +1,40 @@
 exports = module.exports = isEmail
```
Comment (scottgonzalez) [inline index.js]: I know this isn't much shorter, but it does remove the duplication. What do you think?

```js
// Unicode ranges from RFC 3987
//    ucschar        = %xA0-D7FF / %xF900-FDCF / %xFDF0-FFEF
//       / %x10000-1FFFD / %x20000-2FFFD / %x30000-3FFFD
//       / %x40000-4FFFD / %x50000-5FFFD / %x60000-6FFFD
//       / %x70000-7FFFD / %x80000-8FFFD / %x90000-9FFFD
//       / %xA0000-AFFFD / %xB0000-BFFFD / %xC0000-CFFFD
//       / %xD0000-DFFFD / %xE1000-EFFFD
const unicodeRanges = [
  '\u00A0-\uD7FF',
  '\uF900-\uFDCF',
  '\uFDF0-\uFFEF',
  '\u10000-\u1FFFD',
  '\u20000-\u2FFFD',
  '\u30000-\u3FFFD',
  '\u40000-\u4FFFD',
  '\u50000-\u5FFFD',
  '\u60000-\u6FFFD',
  '\u70000-\u7FFFD',
  '\u80000-\u8FFFD',
  '\u90000-\u9FFFD',
  '\uA0000-\uAFFFD',
  '\uB0000-\uBFFFD',
  '\uC0000-\uCFFFD',
  '\uD0000-\uDFFFD',
  '\uE1000-\uEFFFD'
].join('')

const methods = [
  {
    name: 'isEmail',
    characters: unicodeRanges,
    test: () => true
  },
  {
    name: 'isAsciiEmail',
    characters: '',
    test: (parts) => {
      const labels = parts[1].split('.')
      for (const label of labels) {
        if (label.indexOf('xn--') === 0) {
          return false
        }
      }

      return true
    }
  }
].reduce((methods, { characters, name, test }) => {
  const localAddr = new RegExp(`^[a-z0-9.!#$%&'*+/=?^_\`{|}~${characters}-]+$`, 'i')
  const label = `[a-z0-9${characters}](?:[a-z0-9${characters}-]{0,61}[a-z0-9${characters}])?`
  const domain = new RegExp(`^${label}(?:\\.${label})+$`, 'i')

  methods[name] = (string) => {
    const parts = string.split('@')

    if (parts.length !== 2) {
      return false
    }

    if (!localAddr.test(parts[0])) {
      return false
    }

    if (!domain.test(parts[1])) {
      return false
    }

    if (parts[ 0 ].substr(-1) === '.') {
      return false
    }

    return test(parts)
  }

  return methods
}, {})

exports = module.exports = methods.isEmail
exports.isAsciiEmail = methods.isAsciiEmail

exports.isNotEmail = (string) => {
  return !methods.isEmail(string)
}

exports.isNotAsciiEmail = (string) => {
  return !methods.isAsciiEmail(string)
}
```

**Review by scottgonzalez — COMMENTED** [review]
Sorry this took a while to respond to. I'm setting up a new laptop and had to install *everything*.

Overall this looks good. I do like to avoid duplication, so I have a suggested change, but I don't have a strong feeling either way.

File: index.js (hunk 1 above)
Comment (thusoy) [reply index.js]: Works for me! I included this in my latest commit.

**Review by scottgonzalez — APPROVED** [review]

**Comment by scottgonzalez** [comment]
Thanks. This is now on npm as v3.0.1.

v3.0.0 was released as well, but had a typo in the README.

**Comment by thusoy** [comment]
Great, thanks!

**Pull request closed without merging**